### PR TITLE
Bug 1888958: Store secrets in one place and utilize mutex

### PR DESCRIPTION
**Pull request by jcantrill**
### Description
This PR updates the certificate generation code to:
* Add info message why certs are being (re)generated
* Adds a mutex to the cert code to ensure changes are made in a single threaded fashion
* Moves stashing of generated certs to the "master-certs" secret

/assign @syedriko @alanconway 

### Links
* https://bugzilla.redhat.com/show_bug.cgi?id=1888958

### CODE REVIEW

File: scripts/cert_generation.sh [hunk 1]
```diff
@@ -265,3 +299,13 @@ generate_certs 'system.admin'
 generate_certs 'kibana-internal' "$(generate_extensions false false kibana)"
 generate_certs 'elasticsearch' "$(generate_extensions true true $LOG_STORE{,-cluster}{,.${NAMESPACE}.svc}{,.cluster.local})"
 generate_certs 'logging-es' "$(generate_extensions false true $LOG_STORE{,.${NAMESPACE}.svc}{,.cluster.local})"
+
+if [ ! -f "${WORKING_DIR}/kibana-session-secret" ] || [ "$(du -b "${WORKING_DIR}/kibana-session-secret" | cut -f1)" = 0 ] ; then
```
Comment (syedriko) [inline scripts/cert_generation.sh:303]:
```
[ ! -f "${WORKING_DIR}/kibana-session-secret" ] || [ "$(du -b "${WORKING_DIR}/kibana-session-secret" | cut -f1)" = 0 ] 
``` 
is better expressed as
```
[ ! -s "${WORKING_DIR}/kibana-session-secret" ]
```

File: scripts/cert_generation.sh [hunk 2]
```diff
@@ -265,3 +299,13 @@ generate_certs 'system.admin'
...
+  info "Generating kibana session secret"
+  python -c "import string,random;print(''.join(random.choices(string.ascii_letters + string.digits, k=32)))" >  "${WORKING_DIR}/kibana-session-secret"
+fi
+
+if [ -f ${WORKING_DIR}/action.log ] ; then
+  cat ${WORKING_DIR}/action.log
+  rm ${WORKING_DIR}/action.log
+fi
```
Comment (syedriko) [inline scripts/cert_generation.sh:311]: Missing newline here

File: pkg/k8shandler/certificates.go [hunk 3]
```diff
@@ -2,138 +2,81 @@ package k8shandler
...
-		map[string][]byte{
-			"masterca":  utils.GetWorkingDirFileContents("ca.crt"),
-			"masterkey": utils.GetWorkingDirFileContents("ca.key"),
-		})
-
+		secrets,
+	)
 	utils.AddOwnerRefToObject(secret, utils.AsOwner(clusterRequest.Cluster))
 
 	err = clusterRequest.CreateOrUpdateSecret(secret)
```
Comment (syedriko) [inline pkg/k8shandler/certificates.go:49]: this can be just
```
return clusterRequest.CreateOrUpdateSecret(secret)
```

File: pkg/k8shandler/certificates.go [hunk 4]
```diff
@@ -2,138 +2,81 @@ package k8shandler
...
-	return nil
+	return results, nil
 }
 
 //CreateOrUpdateCertificates for a cluster logging instance
 func (clusterRequest *ClusterLoggingRequest) CreateOrUpdateCertificates() (err error) {
-
-	// Pull master signing cert out from secret in logging.Spec.SecretName
-	if err = clusterRequest.readSecrets(); err != nil {
+	mutex.Lock()
```
Comment (syedriko) [inline pkg/k8shandler/certificates.go:71]: It is not quite enough to lock this mutex here, we also need to lock it around accesses to the cert files, in 
```
[localhost cluster-logging-operator]$ ack GetWorkingDirFileContents pkg
pkg/k8shandler/logstore.go
144:		"elasticsearch.key": utils.GetWorkingDirFileContents("elasticsearch.key"),
145:		"elasticsearch.crt": utils.GetWorkingDirFileContents("elasticsearch.crt"),
146:		"logging-es.key":    utils.GetWorkingDirFileContents("logging-es.key"),
147:		"logging-es.crt":    utils.GetWorkingDirFileContents("logging-es.crt"),
148:		"admin-key":         utils.GetWorkingDirFileContents("system.admin.key"),
149:		"admin-cert":        utils.GetWorkingDirFileContents("system.admin.crt"),
150:		"admin-ca":          utils.GetWorkingDirFileContents("ca.crt"),

pkg/k8shandler/curation.go
172:			"ca":       utils.GetWorkingDirFileContents("ca.crt"),
173:			"key":      utils.GetWorkingDirFileContents("system.logging.curator.key"),
174:			"cert":     utils.GetWorkingDirFileContents("system.logging.curator.crt"),
175:			"ops-ca":   utils.GetWorkingDirFileContents("ca.crt"),
176:			"ops-key":  utils.GetWorkingDirFileContents("system.logging.curator.key"),
177:			"ops-cert": utils.GetWorkingDirFileContents("system.logging.curator.crt"),

pkg/k8shandler/visualization.go
146:			"ca":   utils.GetWorkingDirFileContents("ca.crt"),
147:			"key":  utils.GetWorkingDirFileContents("system.logging.kibana.key"),
148:			"cert": utils.GetWorkingDirFileContents("system.logging.kibana.crt"),
160:	sessionSecret = utils.GetWorkingDirFileContents("kibana-session-secret")
170:			"server-key":     utils.GetWorkingDirFileContents("kibana-internal.key"),
171:			"server-cert":    utils.GetWorkingDirFileContents("kibana-internal.crt"),

pkg/k8shandler/fluentd.go
287:			"ca-bundle.crt": utils.GetWorkingDirFileContents("ca.crt"),
288:			"tls.key":       utils.GetWorkingDirFileContents("system.logging.fluentd.key"),
289:			"tls.crt":       utils.GetWorkingDirFileContents("system.logging.fluentd.crt"),
```

File: pkg/k8shandler/certificates.go [hunk 5]
```diff
@@ -2,138 +2,81 @@ package k8shandler
...
+	defer mutex.Unlock()
+	if err = clusterRequest.extractMasterCerts(); err != nil {
+		fmt.Printf("Error %v", err)
 		return
 	}
 
 	scriptsDir := utils.GetScriptsDir()
-	if err = GenerateCertificates(clusterRequest.Cluster.Namespace, scriptsDir, "elasticsearch", utils.DefaultWorkingDir); err != nil {
+	if err = GenerateCertificates(clusterRequest.Cluster.Namespace, scriptsDir, "elasticsearch", utils.GetWorkingDir()); err != nil {
```
Comment (syedriko) [inline pkg/k8shandler/certificates.go:79]: If the script could indicate by its return value that it didn't change anything, we could skip writing the secret and cut down on API server traffic quite a bit.

File: pkg/k8shandler/certificates.go [hunk 6]
```diff
@@ -2,138 +2,81 @@ package k8shandler
...
-		"tls.key":       "system.logging.fluentd.key",
-		"tls.crt":       "system.logging.fluentd.crt",
-	},
-}
+var mutex sync.Mutex
 
-func (clusterRequest *ClusterLoggingRequest) extractSecretToFile(secretName string, key string, toFile string) (err error) {
-	secret, err := clusterRequest.GetSecret(secretName)
+func (clusterRequest *ClusterLoggingRequest) extractMasterCerts() (err error) {
+	secret, err := clusterRequest.GetSecret(constants.MasterCASecretName)
```
Comment (syedriko) [inline pkg/k8shandler/certificates.go:19]: When the new code runs for the first time after an upgrade, it'll dump files "masterca" and "masterkey" on disk and keep them circulating. Just might account for that and skip them altogether.

File: pkg/k8shandler/certificates.go (hunk 3 above)
Comment (syedriko) [inline pkg/k8shandler/certificates.go:49]: this can be just
```
return clusterRequest.CreateOrUpdateSecret(secret)
```

File: pkg/k8shandler/certificates.go [hunk 7]
```diff
@@ -2,138 +2,81 @@ package k8shandler
...
+	return results, nil
 }
 
 //CreateOrUpdateCertificates for a cluster logging instance
 func (clusterRequest *ClusterLoggingRequest) CreateOrUpdateCertificates() (err error) {
-
-	// Pull master signing cert out from secret in logging.Spec.SecretName
-	if err = clusterRequest.readSecrets(); err != nil {
+	mutex.Lock()
+	defer mutex.Unlock()
```
Comment (syedriko) [inline pkg/k8shandler/certificates.go:72]: Checking for 0-length files while populating per-component secrets was one of the bullet points, too.

**Comment by syedriko** [comment]
@jcantrill Added error checking to reading of ES certs:
https://github.com/syedriko/cluster-logging-operator/blob/bz_1888958/pkg/k8shandler/logstore.go#L155

File: scripts/cert_generation.sh (hunk 2 above)
Comment (jcantrill) [reply scripts/cert_generation.sh:311]: Added... but why does it matter if there is a newline after `fi`

Comment (syedriko) [reply scripts/cert_generation.sh:311]: It makes GitHub's diff happy :)

File: pkg/k8shandler/certificates.go (hunk 5 above)
Comment (jcantrill) [reply pkg/k8shandler/certificates.go:79]: updated as we discussed

File: pkg/k8shandler/certificates.go (hunk 7 above)
Comment (jcantrill) [reply pkg/k8shandler/certificates.go]: i believe the changes to the script resolve this

File: pkg/k8shandler/certificates.go [hunk 8]
```diff
@@ -2,163 +2,111 @@ package k8shandler
...
-	for secretKey, certPath := range certs {
-		if err = clusterRequest.extractSecretToFile(secretName, secretKey, certPath); err != nil {
-			if errors.IsNotFound(err) {
-				return nil
-			}
-			return
-		}
+	results := map[string][]byte{}
+	for _, f := range files {
+		results[f.Name()] = utils.GetFileContents(path.Join(workDir, f.Name()))
```
Comment (syedriko) [inline pkg/k8shandler/certificates.go:67]: We could use a check for GetFileContents() failing here.

File: pkg/k8shandler/certificates.go [hunk 9]
```diff
@@ -2,163 +2,111 @@ package k8shandler
...
 	log.V(3).Info("Running script", "script", script, "workDir", workDir, "namespace", namespace, "logStoreName", logStoreName)
 	cmd := exec.Command(script, workDir, namespace, logStoreName)
-	result, err := cmd.Output()
-
-	if log.V(3).Enabled() {
-		log.V(3).Info("cert_generation output", "output", string(result))
-	}
-	if err != nil {
-		log.V(2).Error(err, "Error")
+	out, err := cmd.Output()
```
Comment (syedriko) [inline pkg/k8shandler/certificates.go:105]: Should we check for err != nil and bail right here?

File: pkg/k8shandler/curation.go [hunk 10]
```diff
@@ -164,18 +164,23 @@ func (clusterRequest *ClusterLoggingRequest) createOrUpdateCuratorConfigMap() er
 }
 
 func (clusterRequest *ClusterLoggingRequest) createOrUpdateCuratorSecret() error {
-
-	curatorSecret := NewSecret(
-		"curator",
-		clusterRequest.Cluster.Namespace,
-		map[string][]byte{
+	var secrets = map[string][]byte{}
```
Comment (syedriko) [inline pkg/k8shandler/curation.go:167]: There's no need to allocate a map here
```
var secrets map[string][]byte
```
should do it.

File: pkg/k8shandler/curation.go [hunk 11]
```diff
@@ -164,18 +164,23 @@ func (clusterRequest *ClusterLoggingRequest) createOrUpdateCuratorConfigMap() er
...
+	Syncronize(func() error {
+		secrets = map[string][]byte{
```
Comment (syedriko) [inline pkg/k8shandler/curation.go:169]: We're not checking for possible errors from GetWorkingDirFileContents()

File: pkg/k8shandler/certificates.go (hunk 4 above)
Comment (alanconway) [reply pkg/k8shandler/certificates.go]: This is ok as an interim fix but in the longer term a more maintainable approach is  goroutines+channels (or Communicating Sequential Processes as they were called in the 1960s, sometimes a good idea has to wait for the world to catch up) I can do a short talk about it if that's of interest. The best one-liner from the Go folks is:

    "share memory by communicating, don't communicate by sharing memory".

In summary: don't build mutex walls around thread-safe data structures, only to realize - too late - how threads will end up contending for them. Instead design your process as a set of simple loops with local, contention-free, data; then have them communicate their shared concerns over channels.

Comment (jcantrill) [reply pkg/k8shandler/certificates.go]: @syedriko and I discussed the proposed implementation solely as a fix to the current problem with the understanding we will better design cert generation and syncing when moving to replace the script logic with golang lib

File: pkg/k8shandler/certificates.go (hunk 8 above)
Comment (jcantrill) [reply pkg/k8shandler/certificates.go:67]: The called function logs an error and returns nil.  Adding check for nil and log message. Skipping adding the file if the content is nil

File: pkg/k8shandler/curation.go (hunk 11 above)
Comment (jcantrill) [reply pkg/k8shandler/curation.go:169]: I think this is OK not to check missing here.  We need a bunch of secrets and we should fail and reconcile again.  The cert gen code should regen if something is missing.  If it does not that is a bug in the regen code IMO

Comment (syedriko) [reply pkg/k8shandler/curation.go:169]: But if we don't check we don't fail and stick empty buffers into secrets.

Comment (jcantrill) [reply pkg/k8shandler/curation.go:169]: What is the action we can take in this situation?  What do we believe will change during a subsequent reconciliation that would resolve the issue?  The best we can do is really generate a log message and/or update status.  This particular file is probably moot as we want to remove curation from CLO but  if we are to do anything then it should be applied to the other places where we take similar action

Comment (syedriko) [reply pkg/k8shandler/curation.go:169]: We can abort, refuse to update the secret and potentially make our situation worse. GetWorkingDirFileContents() is not very nuanced in how it reports errors, but there can be a genuine I/O failure that will make GetWorkingDirFileContents() return nil while there's a good cert in the secret as well as on disk. By proceeding we make things worse by wiping out a good cert from a secret. We will recover in time, but this is a disruption we can avoid.

File: scripts/cert_generation.sh [hunk 12]
```diff
@@ -265,3 +299,18 @@ generate_certs 'system.admin'
 generate_certs 'kibana-internal' "$(generate_extensions false false kibana)"
 generate_certs 'elasticsearch' "$(generate_extensions true true $LOG_STORE{,-cluster}{,.${NAMESPACE}.svc}{,.cluster.local})"
 generate_certs 'logging-es' "$(generate_extensions false true $LOG_STORE{,.${NAMESPACE}.svc}{,.cluster.local})"
+
+if [ ! -s "${WORKING_DIR}/kibana-session-secret" ] ; then
+  info "Generating kibana session secret"
+  result=$(python -c "import string,random;print(''.join(random.choice(string.ascii_letters + string.digits) for i in range(32)))")
```
Comment (syedriko) [inline scripts/cert_generation.sh:305]: This achieves almost the same end, with a shorter alphabet of letters, without the Python dependency
```
[root@e33867c18033 cluster-logging-operator]# dd if=/dev/urandom count=1 ibs=16 status=none | hexdump -e '"%02X"'
4306BCE91BC0C99B7AB482ECC3E73047
```

Comment (jcantrill) [reply scripts/cert_generation.sh]: updated to remove python dep
